File: DSA2/hash_table/hash_table.js
// import 'dart:io';

// class Node {
//   int data;
//   Node? next;
//   Node(this.data);
// }

// class HashTable {
//   List<Node?> table;
//   int size;

//   HashTable(this.size) : table = List<Node?>.filled(size, null);

//   int hash(int key) {
//     return key % size;
//   }

//   void insert(int key, int value) {
//     int index = hash(key);
    
//     Node newNode = Node(value);
//     if (table[index] == null) {
//       table[index] = newNode;
//     } else {
//       Node? current = table[index];
//       while (current!.next != null) {
//         current = current.next;
//       }
//       current.next = newNode;
//     }
//   }

//   remove(int key) {
//     int index = hash(key);
//     Node? current = table[index];
//     Node? prev = null;

//     while (current != null && current.data != key) {
//       prev = current;
//       current = current.next;
//     }

//     if (current == null) {
//       print('Key not found');
//       return;
//     }

//     if (prev == null) {
//       table[index] = current.next;
//     } else {
//       prev.next = current.next;
//     }
//   }

//   search(int key) {
//     int index = hash(key);
//     Node? current = table[index];

//     while (current != null) {
//       if (current.data == key) {
//         print('Key found: $key');
//         return;
//       }
//       current = current.next;
//     }
//     print('Key not found: $key');
//   }

//   update(int key, int newValue) {
//     int index = hash(key);
//     Node? current = table[index];

//     while (current != null) {
//       if (current.data == key) {
//         current.data = newValue;
//         print('Key updated: $key to $newValue');
//         return;
//       }
//       current = current.next;
//     }
//     print('Key not found: $key');
//   }

//   bool containsKey(int key) {
//     int index = hash(key);
//     Node? current = table[index];

//     while (current != null) {
//       if (current.data == key) {
//         return true;
//       }
//       current = current.next;
//     }
//     return false;
//   }

//   int get(int key) {
//     int index = hash(key);
//     Node? current = table[index];

//     while (current != null) {
//       if (current.data == key) {
//         return current.data;
//       }
//       current = current.next;
//     }
//     throw Exception('Key not found: $key');
//   }

//   void clear() {
//     for (int i = 0; i < size; i++) {
//       table[i] = null;
//     }
//   }

//   void display() {
//     for (int i = 0; i < size; i++) {
//       stdout.write('Index $i: ');
//       Node? current = table[i];
//       while (current != null) {
//         stdout.write('${current.data} -> ');
//         current = current.next;
//       }
//       print('null');
//     }
//   }
// }

// void main() {
//   HashTable ht = HashTable(10);
//   ht.insert(1, 10);
//   ht.insert(2, 20);
//   ht.insert(11, 30);
//   ht.insert(12, 40);
//   // ht.get(2);
//   ht.update(2, 25);
//   ht.remove(11);
//   ht.search(1);
//   ht.search(11);

//   ht.display();
// }
class HashMap {
    constructor(size = 10) {
        this.map = new Array(size);
    }

    hash(key) {
        let hash =0;
        for(let i=0;i< key.length;i++){
            hash += key.charCodeAt(i);
        }
        return hash % this.map.length;
    }

    set(key, value) {
        let index = this.hash(key);

        if(!this.map[index]) {
            this.map[index] = [];
        }

        for(let pair of this.map[index]) {
            if(pair[0] === key) {
                pair[1] = value;
                return;
            }
        }

        this.map[index].push([key, value]);
    }

    get(key) {
        let index = this._hash(key);

        if(!this.map[index]) return undefined;

        for(let pair of this.map[index]) {
            if(pair[0] === key) {
                return pair[1];
            }
        }

        return undefined;
    }

    remove(key) {
        let index = this._hash(key);

        if(!this.map[index]) return false;

        for(let i=0; i< this.map[index].length; i++) {
            if(this.map[index][i][0] === key) {
                this.map[index].splice(i, 1);
                return true;
            }
        }

        return false;
    }

    display() {
        for(let i=0; i< this.map.length; i++) {
            if(this.map[i]) {
                console.log(i, ":", this.map[i]);
            }
        }
    }
}

const hashMap = new HashMap();
hashMap.set('name', 'John Doe');
hashMap.set('age', 30);
hashMap.set('salary', 45000);
hashMap.set('city', 'New York');

console.log(hashMap.get('name'));
console.log(hashMap.get('age'));
console.log(hashMap.get('city'));

// hashMap.remove('age');

hashMap.display();

File: DSA2/hash_table/hash_table.dart
// class Node<K, V> {
//   K key;
//   V value;
//   Node(this.key, this.value);

//   @override
//   String toString() {
//     return '($key , $value)';
//   }
// }

// class HashTable<K, V> {
//   late List<List<Node<K, V>>?> _table;

//   HashTable([int size = 10]) {
//     _table = List.filled(size, null);
//   }

//   // 🔹 1️⃣ Hash Function: Converts a key to an index
//   int _hash(K key) {
//     return key.hashCode % _table.length;
//   }

//   // 🔹 2️⃣ Insert / Update a key-value pair
//   void set(K key, V value) {
//     int index = _hash(key);
//     _table[index] ??= [];

//     // Check if key already exists (update value)
//     for (var pair in _table[index]!) {
//       if (pair.key == key) {
//         pair = Node(key, value); // Update existing value
//         return;
//       }
//     }

//     // If key is new, add it
//     _table[index]!.add(Node(key, value));
//   }

//   // 🔹 3️⃣ Retrieve a value by key
//   dynamic get(K key) {
//     int index = _hash(key);
//     if (_table[index] == null) return null;

//     for (var pair in _table[index]!) {
//       if (pair.key == key) {
//         return pair.value; // Return value if key matches
//       }
//     }
//     return null; // Key not found
//   }

//   // 🔹 4️⃣ Remove a key-value pair
//   bool remove(K key) {
//     int index = _hash(key);
//     if (_table[index] == null) return false;

//     for (int i = 0; i < _table[index]!.length; i++) {
//       if (_table[index]![i].key == key) {
//         _table[index]!.removeAt(i); // Remove the entry
//         return true;
//       }
//     }
//     return false; // Key not found
//   }

//   // 🔹 5️⃣ Display the entire hash table
//   void display() {
//     for (int i = 0; i < _table.length; i++) {
//       if (_table[i] != null && _table[i]!.isNotEmpty) {
//         print('$i: ${_table[i]}');
//       }
//     }
//   }
// }

// void main() {
//   var hashTable = HashTable<String, String>(87);

//   // Insert values
//   hashTable.set('name', 'John Doe');
//   hashTable.set('age', '30');
//   hashTable.set('salary', '45000');
//   hashTable.set('city', 'New York');

//   // Get values
//   print(hashTable.get('name')); // John Doe
//   print(hashTable.get('age')); // 30
//   print(hashTable.get('city')); // New York

//   // Remove a value
//   // hashTable.remove('age');

//   // Display the entire hash table
//   hashTable.display();
// }

class Node<K, V> {
  K key;
  V value;
  Node(this.key, this.value);
}

class HashTable<K, V> {
  late List<List<Node<K, V>>?> table;
  HashTable([int size = 10]) {
    table = List.filled(size, null);
  }

  hash(K key) {
    return key.hashCode % table.length;

    // var hash = 0;
    // for (int i = 0; i < key.length; i++) {
    //   hash += key.codeUnits(i);
    // }
  }

  set(K key, V val) {
    int index = hash(key);
    table[index] ??= [];

    for (var pair in table[index]!) {
      if (pair.key == key) {
        pair = Node(key, val);
      }
    }
    table[index]!.add(Node(key, val));
  }

  get(K key) {
    int index = hash(key);
    if (table[index] == null) return null;
    for (var pair in table[index]!) {
      if (pair.key == key) {
        return pair.value;
      }
    }
    return false;
  }

  remove(K key) {
    int index = hash(key);
    if (table[index] == null) return null;
    for (var i = 0; i < table[index]!.length; i++) {
      if (table[index]![i].key == key) {
        table[index]!.removeAt(i);
        return;
      }
    }
    return false;
  }

  display() {
    for (var i = 0; i < table.length; i++) {
      if (table[i] != null && table[i]!.isNotEmpty) {
        for (var entry in table[i]!) {
          print(' Key: ${entry.key}, Value: ${entry.value}');
        }
      }
    }
  }
}

void main() {
  var hash = HashTable<String, String>(87);
  hash.set('name', 'John Doe');
  hash.set('age', '30');
  hash.set('salary', '45000');

  hash.set('city', 'New York');
  // hash.set('country', 'USA');
  // hash.set('state', 'NY');
  // hash.set('zip', '10001');
  hash.get('name');

  hash.display();
  print('\n');
  hash.get('age');
  hash.get('city');
  hash.display();
}
